Add troubleshooting SOP for fixing unrecognized certificate authorities (#1919)
openshift/hive: docs/troubleshooting.md

    - [Retrieving stored install logs for a specific cluster provision](#retrieving-stored-install-logs-for-a-specific-cluster-provision)
  - [Deprovision](#deprovision)
  - [HiveAdmission](#hiveadmission)
  - [Regaining access to a cluster after changing the API certificate

> ](#regaining-access-to-a-cluster-after-changing-the-API-certificate)

Author: oops, that's an artifact of copy-pasting

# oc create --raw /apis/admission.hive.openshift.io/v1/dnszones -f config/samples/hiveadmission-review-failure.json -v 8 | jq
```

## Regaining access to a cluster after changing the API certificate

> There are several ways to regain access to a cluster after its API certificate has been changed.  Usually this will manifest as an error message similar to: `x509: certificate signed by unknown authority`.

Reviewer: I think we should frame this from the perspective of having a `ClusterDeployment` in this state with an `Unreachable` condition containing the `x509: certificate signed by unknown authority` error message.

> You will need your new certificate in standard pem format.  It should look something like this:

Reviewer: We need the CA certificate that was used to sign the API certificate.

```suggestion
You will need the certificate authority (CA) for your new API certificate in standard PEM format.  It should look something like this:
```

```sh

> $ cat new-certificate.pem

Reviewer: ```suggestion
$ cat new-ca-certificate.pem
```

```

```
-----BEGIN CERTIFICATE-----
<base64 encoded certificate blob>
-----END CERTIFICATE-----
...
-----BEGIN CERTIFICATE-----
<potentially more base64 encoded certificate blobs>
-----END CERTIFICATE-----
```

### Replacing the API certificate of a managed single cluster with kubeconfig

> In order for a new certificate for a cluster to be accepted by the system, it must be added to the kubeconfig, alongside or replacing any existing certificates. Start by inspecting your current kubeconfig:

Reviewer: We should distinguish that we need the CA certificate to be added to the kubeconfig.

```suggestion
In order for a new API certificate to be accepted by the system, the CA certificate used to sign the new API certificate must be added to the kubeconfig, alongside or replacing any existing certificates. Start by inspecting your current kubeconfig:
```

```sh
$ export MY_KUBECONFIG_NAME=$(oc get cd $CLUSTERNAME -o "jsonpath={.spec.clusterMetadata.adminKubeconfigSecretRef.name}")

> $ oc extract secret/$MY_KUBECONFIG_NAME --keys=kubeconfig --to=- > my-kubeconfig-file

Reviewer 2: ✓
or
```suggestion
$ oc extract secret/$MY_KUBECONFIG_NAME --keys=kubeconfig --to=my-kubeconfig-file
```

$ cat my-kubeconfig-file
```

```yaml
...
clusters:
- cluster:
    certificate-authority-data: <base64 encoded blob>
    server: https://api.<your cluster>.<your domain>:6443
  name: <your cluster>
...
```

The field `certificate-authority-data` is what needs to be updated. When base64 decoded, it looks something like this:

```sh
$ echo <base64 encoded blob> | base64 --decode > decoded-existing-certs.pem
$ cat decoded-existing-certs.pem
```

```
-----BEGIN CERTIFICATE-----
<base64 encoded certificate blob>
-----END CERTIFICATE-----
...
-----BEGIN CERTIFICATE-----
<potentially more base64 encoded certificate blobs>
-----END CERTIFICATE-----
```

> Concatenate your new certificate with the existing decoded certificates:

Reviewer: ```suggestion
Concatenate your new CA certificate with the existing decoded CA certificates:
```

```sh
$ cat decoded-existing-certs.pem new-certificate.pem | openssl base64 -A
```

Copy the contents of this new base64 encoded blob and use it to replace the `certificate-authority-data` field in your kubeconfig. After the kubeconfig file is ready, patch your cluster to use the new kubeconfig:

```sh
oc patch secret $MY_KUBECONFIG_NAME --type='json' -p="[{'op': 'replace', 'path': '/data/kubeconfig', 'value': '$(openssl base64 -A -in my-updated-kubeconfig-file)'}]"
```

> ### Replacing the API certificate globally on the managing cluster with hiveconfig

Reviewer: ```suggestion
### Including a new CA certificate globally on the managing cluster with hiveconfig
```

> Hive can use its own configuration file to regain access to the managed cluster. Use the `hack/set-additional-ca.sh` script to accomplish this:

Reviewer: ```suggestion
Hive can use its own configuration file to configure an additional CA certificate which will be added to kubeconfigs referenced by _ALL_ ClusterDeployments. Use the `hack/set-additional-ca.sh` script to accomplish this:
```

> ```sh

Reviewer 3: Please note that the command `oc patch` needs to be updated based on this interaction in the original issue:
https://github.com/openshift/hive/issues/1900#issuecomment-1333929786

This would need confirmation from your test automation, but I believe the command should be:

```sh
oc patch secret $MY_KUBECONFIG_NAME --type='json' -p="[{'op': 'replace', 'path': '/data/kubeconfig', 'value': '$(openssl base64 -A -in my-updated-kubeconfig-file)'},{'op': 'replace', 'path': '/data/raw-kubeconfig', 'value': '$(openssl base64 -A -in my-updated-kubeconfig-file)'}]"
```

# oc create --raw /apis/admission.hive.openshift.io/v1/dnszones -f config/samples/hiveadmission-review-failure.json -v 8 | jq
```

## Regaining access to a cluster after changing the API certificate

There are several ways to regain access to a cluster after its API certificate has been changed.  Usually this will manifest as an error message similar to: `x509: certificate signed by unknown authority`.

### Replacing the API certificate of a managed single cluster with kubeconfig

In order for a new certificate for a cluster to be accepted by the system, it must be added to the kubeconfig, alongside or replacing any existing certificates. Start by inspecting your current kubeconfig:

```sh
$ export MY_KUBECONFIG_NAME=$(oc get cd $CLUSTERNAME -o "jsonpath={.spec.clusterMetadata.adminKubeconfigSecretRef.name}")

> $ oc get secret $MY_KUBECONFIG_NAME -o "jsonpath={.data.kubeconfig}" | base64 --decode > my-kubeconfig-file

Reviewer 2: FYI there's an `oc extract` command that looks slightly less magical. Exercise for the reader to figure out how it works :)

$ cat my-kubeconfig-file
```

```yaml
...
clusters:
- cluster:
    certificate-authority-data: <base64 encoded blob>
    server: https://api.<your cluster>.<your domain>:6443
  name: <your cluster>
...
```

The field `certificate-authority-data` is what needs to be updated. When base64 decoded, it looks something like this:

```sh
$ echo <base64 encoded blob> | base64 --decode > decoded-existing-certs.pem
$ cat decoded-existing-certs.pem
```

```
-----BEGIN CERTIFICATE-----
<base64 encoded certificate blob>
-----END CERTIFICATE-----
...
-----BEGIN CERTIFICATE-----
<potentially more base64 encoded certificate blobs>
-----END CERTIFICATE-----
```

Concatenate your new certificate with the existing decoded certificates:

```sh

> $ cat decoded-existing-certs.pem <your new certificate full chain>.pem | openssl base64 -A > encoded-new-cert-bundle.pem

Reviewer 2: nit: Calling this a .pem file is technically inaccurate

```

Copy the contents of this new base64 encoded blob and use it to replace the `certificate-authority-data` field in your kubeconfig. After the kubeconfig file is ready, patch your cluster to use the new kubeconfig:

```sh
oc patch secret $MY_KUBECONFIG_NAME --type='json' -p="[{'op': 'replace', 'path': '/data/kubeconfig', 'value': '$(openssl base64 -A -in my-updated-kubeconfig-file)'}]"
```

### Replacing the API certificate globally on the managing cluster with hiveconfig

Hive can use its own configuration file to regain access to the managed cluster. Use the `hack/set-additional-ca.sh` script like so: 

```sh

> $ hack/set-additional-ca.sh <file location of your chain.pem>

Reviewer 2: unclear whether this is supposed to be the encoded (base64) or decoded (PEM) file.
